File: id-card-toolkit-ios-sdk-v3.1.6/id-card-toolkit-ios-sdk-v3.1.6/samples/IDCardToolkitObjCApp/IDCardToolkitObjCApp/ViewController/UnblockPINViewController.m
//
//  UnblockPINViewController.m
//  
//
//  Created by Federal Authority For Identity and Citizenship  on 5/29/17.
//  Copyright © 2017 Federal Authority For Identity and Citizenship. All rights reserved.
//

#import "UnblockPINViewController.h"

@interface UnblockPINViewController ()
@property(nonatomic,strong)Utils * utils;
@end

@implementation UnblockPINViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.pintext.keyboardType=UIKeyboardTypeNumberPad;
    self.pintext.keyboardAppearance = UIKeyboardAppearanceLight;
    UIToolbar* enterPinNumberToolbar = [[UIToolbar alloc]initWithFrame:CGRectMake(0, 0, 320, 50)];
    enterPinNumberToolbar.barStyle = UIBarStyleBlackTranslucent;
    enterPinNumberToolbar.items = @[[[UIBarButtonItem alloc]initWithTitle:@"Cancel" style:UIBarButtonItemStyleDone target:self action:@selector(enterPinCancelNumberPad)],
                                    [[UIBarButtonItem alloc]initWithBarButtonSystemItem:UIBarButtonSystemItemFlexibleSpace target:nil action:nil],
                                    [[UIBarButtonItem alloc]initWithTitle:@"Done" style:UIBarButtonItemStyleDone target:self action:@selector(enterPinDoneWithNumberPad)]];
    [enterPinNumberToolbar sizeToFit];
    self.pintext.inputAccessoryView = enterPinNumberToolbar;
    
    self.title=@"Unblock Pin";
    self.navigationItem.hidesBackButton = YES;
    self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithImage:[[UIImage imageNamed:@"backbutton"] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal]
                                                                             style:UIBarButtonItemStylePlain
                                                                            target:self
                                                                            action:@selector(goBack)];
    
    self.unblockPinButton.layer.cornerRadius=22.0;
    [self.unblockPinButton.layer setShadowOffset:CGSizeMake(5, 5)];
    [self.unblockPinButton.layer setShadowColor:[[UIColor lightGrayColor] CGColor]];
    [self.unblockPinButton.layer setShadowOpacity:0.5];
  //  self.unblockPinButton.layer.borderWidth=1.0;
    
    AppDelegate *aDelegate = (AppDelegate *)[UIApplication sharedApplication].delegate;
    self.model = aDelegate.model;
    self.utils=aDelegate.utils;
    
    // Do any additional setup after loading the view.
}
#pragma mark viewWillAppear

-(void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:YES];
    
    self.pintext.text=@"";
    [self.pintext resignFirstResponder];
}
#pragma mark EIDA Toolkit calling functions

- (IBAction)unblockPinButtonAction:(id)sender {
    
 if ([self.pintext.text length]==0) {
    [AlertView showAlertTitle:ALERT withMessage:ENTER_PIN onView:self];
 }// if
 else {
    [self.utils ShowProgressBar:UNBLOCKPIN andView:self.view];
    [self.pintext resignFirstResponder];
     NSString *requestId=[Utils generateSecureKey];
     dispatch_async(dispatch_get_global_queue( DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^(void){
    @try {
        CardReader *cardreader  = [self.model.cardReaderSharedArray objectAtIndex:0];
        FingerData *selectedFingerData =[self.model.selectedFingerSharedArray objectAtIndex:0];
        DataProtectionKey *data_protection_key=self.model.dataprotectedShared;
        
        NSString *request_handle =  [cardreader prepareRequest:requestId];
        if (request_handle.length==0 || [request_handle isEqualToString:@""] || [request_handle isEqualToString:@"(null)"] ) {
            [AlertView showAlertTitle:ALERT withMessage:REQUEST_HANDLE_EMPTY onView:self];
        }//if
        else {
            NSString *encodePin=[Utils setEncrytion:request_handle data:self.pintext.text publickey:[data_protection_key getPublicKey] keylength:[data_protection_key getKeyLength]];
            if (encodePin.length==0 || [encodePin isEqualToString:@""] || [encodePin isEqualToString:@"(null)"] ) {
                [AlertView showAlertTitle:ALERT withMessage:ENCODE_PIN_EMPTY onView:self];
            }//if
            else {
                 ToolkitResponse *response = [cardreader unblockPin:encodePin fingerData:selectedFingerData sensorTimeout:SENSOR_TIME_OUT];
                dispatch_async(dispatch_get_main_queue(), ^{
                    NSString *xmlString =[response getXmlString];
                    if (![xmlString isEqualToString:@""] && xmlString.length>0) {
                        NSString *errorMsg =  [self.utils validateToolkitResponse:requestId xmlstring:xmlString];
                        if (![errorMsg isEqualToString:@""] && errorMsg.length>0) {
                            [AlertView showAlertTitle:ALERT withMessage:errorMsg onView:self];
                        }//if
                        else {
                            NSString *xmlString =[response getXmlString];
                            [AlertView showAlertTitle:ALERT withMessage:[NSString stringWithFormat:@"%@\n\n%@",CARD_UNBLOCK_SUCCESS,xmlString] onView:self];
                        }//else
                    }//if
                    else {
                        ResponseStatus *status = [response getResponseStatus];
                        if (status==0) {
                            [AlertView showAlertTitle:ALERT withMessage:CARD_UNBLOCK_SUCCESS onView:self];
                        }//if
                    }//else
                });// update UI main queue
            }//else
          }//else
       }//try
      @catch (NSException *exception) {
            NSString *exe =[NSString stringWithFormat:@"%@\n%@\n%@",exception.name,exception.description,exception.userInfo];
            [AlertView showAlertTitle:ALERT withMessage:exe onView:self];
      }// catch
        @finally {
             dispatch_async(dispatch_get_main_queue(), ^{
                 self.pintext.text=@"";
                 [self.utils DismissProgressBar];
             });// update UI main queue
          }// finally
       });// background queue
    }//else
}
#pragma mark Updating UI Methods

-(void)enterPinCancelNumberPad {
    [self.pintext resignFirstResponder];
    self.pintext.text = @"";
}
-(void)enterPinDoneWithNumberPad {
    [self.pintext resignFirstResponder];
}
- (void)goBack {
    [self.navigationController popViewControllerAnimated:YES];
}
#pragma mark didReceiveMemoryWarning
- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/


@end

File: id-card-toolkit-ios-sdk-v3.1.6/id-card-toolkit-ios-sdk-v3.1.6/samples/ICASwiftApp/ICASwiftApp/UnblockPINViewController.swift
//
//  UnblockPINViewController.swift
//  ICASwiftApp
//
//  Created by Digital Trust Tech on 09/05/22.
//

import UIKit

class UnblockPINViewController: UIViewController {
    
    @IBOutlet weak var pintext: UITextField!
    @IBOutlet var unblockPinButton: UIButton!
    @IBOutlet weak var view_Roundedcorners: UIView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.view_Roundedcorners.cornered_View([.layerMaxXMaxYCorner, .layerMinXMaxYCorner], radius: 20, borderColor: .red, borderWidth: 0)
        self.pintext.keyboardType = .numberPad
        self.pintext.keyboardAppearance = .light
        let enterPinNumberToolbar = UIToolbar(frame: CGRect(x: CGFloat(0), y: CGFloat(0), width: CGFloat(320), height: CGFloat(50)))
        enterPinNumberToolbar.barStyle = .blackTranslucent
        enterPinNumberToolbar.items = [UIBarButtonItem(title: "Cancel", style: .done, target: self, action: #selector(enterPinCancelNumberPad)), UIBarButtonItem(barButtonSystemItem: .flexibleSpace, target: nil, action: nil), UIBarButtonItem(title: "Done", style: .done, target: self, action: #selector(enterPinDoneWithNumberPad))]
        enterPinNumberToolbar.sizeToFit()
        self.pintext.inputAccessoryView=enterPinNumberToolbar

        unblockPinButton.layer.cornerRadius = 20.0
        unblockPinButton.layer.shadowOffset = CGSize(width: 5.0, height: 5.0)
        unblockPinButton.layer.shadowColor = UIColor.lightGray.cgColor
        unblockPinButton.layer.shadowOpacity = 0.5
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(true)
        
        pintext.text=""
        pintext.resignFirstResponder()
    }
    
    @IBAction func btn_LoadFingerData(_ sender: Any) {
        let vc = storyboard?.instantiateViewController(withIdentifier: "FingerDataViewController") as! FingerDataViewController
        present(vc, animated: true, completion: nil)
    }
    
    @IBAction func unblockPinButtonAction(_ sender: Any) {
        
        if toolkit == nil  {
            self.view.makeToast(Constants.toolkitNotInitilized)
            return
        }//if
        else if cardreader == nil {
            self.view.makeToast(Constants.cardreaderNotConnect)
            return
        }//else if
        else if selectedFingerData == nil {
            self.view.makeToast(Constants.fingerDataNotFound)
            return
        }//else if
        
        self.pintext.resignFirstResponder()
        if self.pintext.text?.count==0 {
            AlertView.showAlertTitle(Constants.Alert, withMessage: Constants.EnterPin, onView: self)
        }// if
        else {
            utils.showProgressBar(Constants.UnblockPin, andView: self.view)
            let requestId = Utils.generateSecureKey()
            DispatchQueue.global(qos: .background).async {
                do {
                    let request_handle = try cardreader?.prepareRequest(requestId!)
                    if (request_handle?.isEmpty == true) || (request_handle == "") || (request_handle == "(null)") {
                        AlertView.showAlertTitle(Constants.Alert, withMessage:Constants.Request_Handle_Empty, onView: self)
                    }//if
                    else {
                        let encodePin:String? = Utils.setEncrytion(request_handle, data: self.pintext.text!,publickey: dataProtectionKey.getPublicKey(), keylength: Int32(dataProtectionKey.getKeyLength()))
                        if (encodePin?.isEmpty)! || (encodePin == "") || (encodePin == "(null)") {
                            AlertView.showAlertTitle(Constants.Alert, withMessage:Constants.Encode_Pin_Empty, onView: self)
                        }//if
                        else {
                            response = try cardreader?.unblockPin(encodePin!, fingerData: selectedFingerData, sensorTimeout: Int32(Constants.SensorTimeOut))
                            DispatchQueue.main.async {
                                let xmlString = response.getXmlString()
                                if (xmlString != "") && ((xmlString.count ) > 0) {
                                    let errorMsg: String = utils.validateToolkitResponse(requestId, xmlstring: xmlString)
                                    if (errorMsg != "") && (errorMsg.count ) > 0 {
                                        AlertView.showAlertTitle(Constants.Alert, withMessage: errorMsg, onView: self)
                                    }//if
                                    else {
                                        let xmlString = response.getXmlString()
                                        AlertView.showAlertTitle(Constants.Alert, withMessage: "\(Constants.CardUnblockedSuccess)\n\n \(xmlString)", onView: self)
                                    }//else
                                }//if
                                else {
                                    let status = response.getResponseStatus()
                                    AlertView.showAlertTitle(Constants.Alert, withMessage: status, onView: self)
                                }//else
                            }// update UI main queue
                        }//else
                    }//else
                    DispatchQueue.main.async {
                        utils.dismissProgressBar()
                        self.pintext.text=""
                    }// update UI main queue
                }//do
                catch let error as NSError{
                    DispatchQueue.main.async {
                        let err="\(error.domain)\n\(error.code)\n\(error.userInfo)"
                        AlertView.showAlertTitle(Constants.Alert, withMessage:err,onView: self)
                        utils.dismissProgressBar()
                        self.pintext.text=""
                    }// update UI main queue
                }//catch
            }//background queue
        }//else
    }
    
    @objc func enterPinCancelNumberPad() {
        pintext.resignFirstResponder()
        pintext.text=""
    }
    
    @objc func enterPinDoneWithNumberPad() {
        pintext.resignFirstResponder()
    }
    @objc func goBack() {
        navigationController?.popViewController(animated: true)
    }

    @IBAction func btn_BackAction(_ sender: Any) {
        navigationController?.popViewController(animated: true)
    }
    
}
